community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OffHeapPageLock.java

```diff
@@ -0,0 +1,354 @@
+     * if no write or exclusive lock overlapped with the critical section of the optimistic read lock represented by
+     * the stamp.
+     *
+     * @param stamp The stamp of the optimistic read lock.
+     * @return {@code true} if the optimistic read lock was valid, {@code false} otherwise.
+     */
+    public static boolean validateReadLock( long address, long stamp )
+    {
+        UnsafeUtil.loadFence();
+        return (getState( address ) & CHK_MASK) == stamp;
+    }
+
+    public static boolean isModified( long address )
+    {
+        return (getState( address ) & MOD_MASK) == MOD_MASK;
+    }
+
+    public static boolean isExclusivelyLocked( long address )
+    {
+        return (getState( address ) & EXL_MASK) == EXL_MASK;
+    }
+
+    /**
+     * Try taking a concurrent write lock. Multiple write locks can be held at the same time. Write locks will
+     * invalidate any optimistic read lock that overlaps with them, and write locks will make any attempt at grabbing
+     * an exclusive lock fail. If an exclusive lock is currently held, then the attempt to take a write lock will fail.
+     * <p>
+     * Write locks must be paired with a corresponding {@link #unlockWrite(long)}.
+     *
+     * @return {@code true} if the write lock was taken, {@code false} otherwise.
+     */
+    public static boolean tryWriteLock( long address )
+    {
+        long s, n;
+        for (; ; )
+        {
+            s = getState( address );
+            boolean unwritablyLocked = (s & EXL_MASK) != 0;
+            boolean writeCountOverflow = (s & CNT_MASK) == CNT_MASK;
+
+            // bitwise-OR to reduce branching and allow more ILP
+            if ( unwritablyLocked | writeCountOverflow )
+            {
+                return failWriteLock( s, writeCountOverflow );
+            }
+
+            n = s + CNT_UNIT | MOD_MASK;
+            if ( compareAndSetState( address, s, n ) )
+            {
+                UnsafeUtil.storeFence();
+                return true;
+            }
+        }
+    }
+
+    private static boolean failWriteLock( long s, boolean writeCountOverflow )
+    {
+        if ( writeCountOverflow )
+        {
+            throwWriteLockOverflow( s );
+        }
+        // Otherwise it was exclusively locked
+        return false;
+    }
+
+    private static long throwWriteLockOverflow( long s )
+    {
+        throw new IllegalMonitorStateException( "Write lock counter overflow: " + describeState( s ) );
+    }
+
+    /**
+     * Release a write lock taking with {@link #tryWriteLock(long)}.
+     */
+    public static void unlockWrite( long address )
+    {
+        long s, n;
+        do
+        {
+            s = getState( address );
+            if ( (s & CNT_MASK) == 0 )
+            {
+                throwUnmatchedUnlockWrite( s );
+            }
+            n = nextSeq( s ) - CNT_UNIT;
+        }
+        while ( !compareAndSetState( address, s, n ) );
+    }
```

Comment:
There is in principle no harm in coarsening a critical section. It's a _roach motel_; you can move in, but you can never move out.